### PR TITLE
Tabular model

### PR DESCRIPTION
**Pull request by nden**
This is an implementation of a lookup table model with interpolation done by `scipy's interpn` funciton.
@pllim Please review if this will work for you.

### CODE REVIEW

**Comment by pllim** [comment]
Thanks, @nden ! I'll have a look when I get the chance. :smile:

**Comment by pllim** [comment]
@nden, perhaps add an example on how one might use this in fitting? I imagine this would generate data points for the fitter, not as the model; e.g., someone would use a Gaussian model and a fitter to try to fit something to data stored in this model.

**Comment by pllim** [comment]
@nden, that way it is currently implemented, I am not sure how to subclass it for `pysynphot`.

```python
class Tabular1D(Tabular):
    pass
```

```python
>>> tab = Tabular1D(data)
TypeError: Can't instantiate abstract class Tabular2D with abstract methods lookup_table
```

**Comment by nden** [comment]
@pllim Look at the example for `tabular_model`. You need to create first the class with an appropriate table (1D or 2D, ...), e.g.

Tabular1D = tabular_model(my_1d_table)

Then you can subclass `Tabular1D` or initialize it (e.g. with a different 1D table).

For discussion on why this is done in such a way see spacetelescope/gwcs#28

**Comment by nden** [comment]
I'm not sure I understand what example you have in mind. It seems what you are describing is simply fitting a gaussian model to a data array.

**Comment by pllim** [comment]
As for the example, yes, basically as you said. So far, all the fitting example use the models the "conventional" way. So such an example might answer the question of, "But how does the tabular model fit in fitting?"

**Comment by nden** [comment]
@pllim If you have such example please add it.

**Comment by pllim** [comment]
Sure. I'll play around with this somemore.

**Comment by pllim** [comment]
Update: I am just coming back to this after many other distractions... Please give me a bit more time...

**Comment by pllim** [comment]
@nden , I looked at the discussions you had with @embray and the examples. I find this a bit clunky for the usage I have in mind. And also a bit confused. See below.

```python
LookupTable = models.tabular_model(np.arange(10))
model = LookupTable(points=points, lookup_table=values)
```

In the example above (from your test), it appears `np.arange(10)` is just a dummy input to set the dimensions. I don't feel good about making fake data just to be able to create a class out of it. Why can't I just pass in the dimension (shape) directly?

Also, for my usage, I need to inherit it and overwrite the `__init__()` to handle units and negative flux. In its current form, I need to do something like this:

```python
from astropy.modeling.models import tabular_model

# This just does not make sense to me
Tabular1D = tabular_model(fake_data) 

class Empirical1D(Tabular1D):
    """Source spectrum from data files."""
    def __init__(self, ...):
        super(Empirical1D, self).__init__(...)
        # Use points and lookup values from file
        # Extra processing
```

With the intended usage something like this (simplified):

```python
tab = QTable.read('myspectrum.fits', format='fits')
wave, flux = tab['WAVELENGTH'], tab['FLUX']  # Quantity
model_instance = Empirical1D(wave, flux)
user_flux = model_instance(user_wave)
```

File: astropy/modeling/tabular.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        Parameters
+        ----------
+        inputs : list of scalars or ndarrays
+            Input coordinates. The number of inputs must be equal
+            to the dimensions of the lookup table.
+        """
+        inputs = np.array(inputs[: self.n_inputs]).T
+        if not has_scipy:
+            raise ImportError("This model requires scipy >= v0.14")
+        return interpn(self.points, self.lookup_table, inputs,
```
Comment (pllim) [inline astropy/modeling/tabular.py]: When I try to use this with `points` in descending order (which PySynphot and `interp1d` both support), I get:

```python
ValueError: The points in dimension 0 must be strictly ascending
```

Is there a way to get around this without having to re-order the data myself? I could overwrite my subclass' `evaluate()` using `interp1d` but that kind of defeats the purpose of subclassing.

Comment (pllim) [inline astropy/modeling/tabular.py]: Update: I got around this by explicitly flipping the points if they are not of ascending order *before* passing them into parent constructor.

**Comment by pllim** [comment]
Overall, it works for my 1D cases. I was able to work around the quirks. I wonder if it would be convenient to also provide the following in your PR:

```python
Tabular1D = tabular_model([0, 0])
"""One dimensional tabular model."""

Tabular2D = tabular_model([[0, 0], [0, 0]])
"""Two dimensional tabular model."""
``` 

In any case, I would rather this merge sooner than later, so we both can move forward. Thanks!

File: astropy/modeling/tabular.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        Interpolated values at input coordinates.
+
+    Raises
+    ------
+    ImportError
+        Scipy is not installed.
+
+    Notes
+    -----
+    Uses ``scipy.interpolate.interpn``
```
Comment (pllim) [inline astropy/modeling/tabular.py]: Maybe use intersphinx here to link to API doc?

**Comment by pllim** [comment]
I don't see any other people with concerns and I want to start using this, so I'll merge this in an hour or two unless I hear otherwise. We can follow-up with separate PR later if needed.

**Comment by nden** [comment]
Well, give me a chance to get to my desk. Just coming back from vacation.

**Comment by nden** [comment]
@pllim Take a look at the latest commit. To create a class now you need to pass the dimension of the table. I also created `Tabular1D` and `Tabular2D`. Feel free to merge when tests pass and if this works for you.

**Comment by pllim** [comment]
@nden -- Is it possible to add a `bounding_box` like the rest of the models? Something like this might work:

```python
class Tabular(...):
    @property
    def bounding_box(self):
        """Tuple defining the default ``bounding_box`` limits,
        ``(points_low, points_high)``.

        Examples
        -------------
        >>> from astropy.modeling.models import Tabular1D, Tabular2D
        >>> t1 = Tabular1D([1,2,3], [10, 20, 30])
        >>> t1.bounding_box
        (1, 3)
        >>> t2 = Tabular2D([[1,2,3],[2,3,4]], [[10,20,30],[20,30,40]])
        >>> t2.bounding_box
        ((2, 4), (1, 3))

        """
        bbox = [(min(p), max(p)) for p in self.points][::-1]
        if len(bbox) == 1:
            bbox = bbox[0]
        return tuple(bbox)
```

This follows the following documentation in `core.py`:

```
    @property
    def bounding_box(self):
        """
        A `tuple` of length `n_inputs` defining the bounding box limits, or
        `None` for no bounding box.

        The default limits are given by a ``bounding_box`` property or method
        defined in the class body of a specific model.  If not defined then
        this property just raises `NotImplementedError` by default (but may be
        assigned a custom value by a user).  ``bounding_box`` can be set
        manually to an array-like object of shape ``(model.n_inputs, 2)``. For
        further usage, see :ref:`bounding-boxes`

        The limits are ordered according to the `numpy` indexing
        convention, and are the reverse of the model input order,
        e.g. for inputs ``('x', 'y', 'z')``, ``bounding_box`` is defined:

        * for 1D: ``(x_low, x_high)``
        * for 2D: ``((y_low, y_high), (x_low, x_high))``
        * for 3D: ``((z_low, z_high), (y_low, y_high), (x_low, x_high))``

        Examples
        --------

        Setting the ``bounding_box`` limits for a 1D and 2D model:

        >>> from astropy.modeling.models import Gaussian1D, Gaussian2D
        >>> model_1d = Gaussian1D()
        >>> model_2d = Gaussian2D(x_stddev=1, y_stddev=1)
        >>> model_1d.bounding_box = (-5, 5)
        >>> model_2d.bounding_box = ((-6, 6), (-5, 5))

        Setting the bounding_box limits for a user-defined 3D `custom_model`:

        >>> from astropy.modeling.models import custom_model
        >>> def const3d(x, y, z, amp=1):
        ...    return amp
        ...
        >>> Const3D = custom_model(const3d)
        >>> model_3d = Const3D()
        >>> model_3d.bounding_box = ((-6, 6), (-5, 5), (-4, 4))

        To reset ``bounding_box`` to its default limits just delete the
        user-defined value--this will reset it back to the default defined
        on the class:

        >>> del model_1d.bounding_box

        To disable the bounding box entirely (including the default),
        set ``bounding_box`` to `None`:

        >>> model_1d.bounding_box = None
        >>> model_1d.bounding_box  # doctest: +IGNORE_EXCEPTION_DETAIL
        Traceback (most recent call last):
          File "<stdin>", line 1, in <module>
          File "astropy\modeling\core.py", line 980, in bounding_box
            "No bounding box is defined for this model (note: the "
        NotImplementedError: No bounding box is defined for this model (note:
        the bounding box was explicitly disabled for this model; use `del
        model.bounding_box` to restore the default bounding box, if one is
        defined for this model).
        """
```

Even if it is not included here, I have to define it for my use anyway, so might as well, right?

**Comment by pllim** [comment]
@nden -- Also maybe expose `points` and `lookup_table` in `str()` and `repr()` like this? Currently, those methods are not very informational for tabular models.

```python
class Tabular(...):
    def __repr__(self):
        return self._format_repr(kwargs={
            'points': self.points, 'lookup_table': self.lookup_table})

    def __str__(self):
        return self._format_str(keywords=[
            ('points', self.points), ('lookup_table', self.lookup_table)])
```

This would result in such printouts:

```python
>>> t1 = Tabular1D([1,2,3], [10, 20, 30])
>>> str(t1)
"Model: T1\nInputs: ('x0',)\nOutputs: ('y',)\nModel set size: 1\npoints: ([1, 2, 3],)\nlookup_table: [10 20 30]\nParameters:"
>>> repr(t1)
'<T1(points=([1, 2, 3],), lookup_table=array([10, 20, 30]))>'
>>> print(t1)
Model: T1
Inputs: ('x0',)
Outputs: ('y',)
Model set size: 1
points: ([1, 2, 3],)
lookup_table: [10 20 30]
Parameters:
```

**Comment by pllim** [comment]
@nden -- Also, I can't image how Tabular would work for `n_models > 1`, so we should just explicitly disable it like this:

```python
class Tabular(...):
    def __init__(self, ..., **kwargs):
        n_models = kwargs.get('n_models', 1)
        if n_models > 1:
            raise NotImplementedError('Only n_models=1 is supported')

        super(...)
```

File: astropy/modeling/tabular.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+                                 "{0} dimensions".format(self.lookup_table.ndim))
+            self.lookup_table = lookup_table
+        if points is None:
+            self._points = tuple(np.arange(x, dtype=np.float)
+                                 for x in self.lookup_table.shape)
+        else:
+            if self.lookup_table.ndim == 1:
+                self._points = (points,)
+            else:
+                self._points = points
```
Comment (pllim) [inline astropy/modeling/tabular.py]: I think right after this `if-else` block, we have to add a line like this:
```python
self._points = np.squeeze(self._points)
```

This is to produce something like:

```python
>>> t1 = Tabular1D([1,2,3], [10, 20, 30])
>>> t1.points   # Currently: ([1, 2, 3],)
array([1, 2, 3])
>>> t2 = Tabular2D([[1,2,3],[2,3,4]], [[10,20,30],[20,30,40]])
>>> t2.points   # Currently: [[1, 2, 3], [2, 3, 4]]
array([[1, 2, 3],
       [2, 3, 4]])
```

This is to be consistent with `lookup_table`:

```python
>>> t1.lookup_table
array([10, 20, 30])
>>> t2.lookup_table
array([[10, 20, 30],
       [20, 30, 40]])
```

**Comment by pllim** [comment]
Last but not least, I thought about the fitting example that I mentioned earlier. It would seem repetitive to do it because basically it is just replacing `x` and `y` in [Simple 1-D model fitting](http://docs.astropy.org/en/latest/modeling/index.html#simple-1-d-model-fitting) with `t1.points` and `t1.lookup_table`. I guess we can leave it out for now until someone actually asks about it.

Thank you for this PR!

File: astropy/modeling/tabular.py (hunk 2 above)
Comment (pllim) [inline astropy/modeling/tabular.py]: @nden -- Ping. Yes or no to intersphinx linking here?

File: astropy/modeling/tabular.py (hunk 3 above)
Comment (pllim) [inline astropy/modeling/tabular.py]: Update: It appears that my suggestion above breaks `evaluate()`...

Comment (pllim) [inline astropy/modeling/tabular.py]: How about like this instead?

```python
class Tabular(...):
    @property
    def points(self):
        return np.squeeze(self._points)

    def evaluate(self, *inputs):
        ...
        return interpn(self._points, ...)
```

File: astropy/modeling/tests/test_models.py [hunk 4]
```diff
@@ -23,14 +23,18 @@
...
 
 try:
+    import scipy
     from scipy import optimize  # pylint: disable=W0611
     HAS_SCIPY = True
 except ImportError:
     HAS_SCIPY = False
 
+if HAS_SCIPY and not minversion(scipy, "0.14"):
+    HAS_SCIPY = False
```
Comment (pllim) [inline astropy/modeling/tests/test_models.py]: Just an observation: This skips all the other existing tests that currently requires SciPy but not necessarily v0.14 or later. Perhaps it is more accurate to make a new flag called `HAS_SCIPY_14` (or something like that) and then explicitly use it for `Tabular` tests only?

File: astropy/modeling/tests/test_models.py [hunk 5]
```diff
@@ -516,3 +520,55 @@ def test_voigt_model():
 def test_model_instance_repr():
     m = models.Gaussian1D(1, 2, 3)
     assert repr(m) == '<Gaussian1D(amplitude=1.0, mean=2.0, stddev=3.0)>'
+
+
+@pytest.mark.skipif("not HAS_SCIPY")
+def test_tabular_interp_1d():
+    points = np.arange(0, 5)
+    values = [1., 10, 2, 45, -3]
+    LookupTable = models.tabular_model(1)
```
Comment (pllim) [inline astropy/modeling/tests/test_models.py]: Why not just use `Tabular1D` that you defined here?

File: astropy/modeling/tests/test_models.py [hunk 6]
```diff
@@ -516,3 +520,55 @@ def test_voigt_model():
...
+    table = np.array([[-0.04614432, -0.02512547, -0.00619557, 0.0144165 , 0.0297525 ],
+       [-0.04510594, -0.03183369, -0.01118008, 0.01201388, 0.02496205],
+       [-0.05464094, -0.02804499, -0.00960086, 0.01134333, 0.02284104],
+       [-0.04879338, -0.02539565, -0.00440462, 0.01795145, 0.02122417],
+       [-0.03637372, -0.01630025, -0.00157902, 0.01649774, 0.01952131]])
+
+    points = (np.arange(0, 5), np.arange(0, 5))
+
+    xnew = np.array([0., .7, 1.4, 2.1, 3.9])
+    LookupTable = models.tabular_model(2)
```
Comment (pllim) [inline astropy/modeling/tests/test_models.py]: Why not just use `Tabular2D` that you defined here?

File: astropy/modeling/tests/test_models.py [hunk 7]
```diff
@@ -516,3 +520,55 @@ def test_voigt_model():
...
+    model = LookupTable(points, table)
+    znew = model(xnew, xnew)
+    result = np.array([-0.04614432, -0.03450009, -0.02241028, -0.0069727 ,  0.01938675])
+    utils.assert_allclose(znew, result, atol=10**-7)
+    with pytest.raises(ValueError):
+        model = LookupTable(points=([1.2, 2.3], [1.2, 6.7], [3, 4]))
+
+
+@pytest.mark.skipif("not HAS_SCIPY")
+def test_tabular():
```
Comment (pllim) [inline astropy/modeling/tests/test_models.py]: This contains repetition from `test_tabular_interp_1d()` above. Is it really necessary? Can this become apart of `test_tabular_interp_1d()` instead?

File: astropy/modeling/tests/test_models.py (hunk 4 above)
Comment (pllim) [inline astropy/modeling/tests/test_models.py]: Or maybe just somehow use `tabular.has_scipy` specifically for `Tabular` tests without modifying `HAS_SCIPY` here.

File: astropy/modeling/tests/test_models.py (hunk 6 above)
Comment (pllim) [inline astropy/modeling/tests/test_models.py]: Although, perhaps it is good to call `tabular_model` without `name=` to test that part of the logic...

File: astropy/modeling/tabular.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+                raise ValueError("Expected grid points in "
+                                 "{0} directions, got {1}".format(self.lookup_table.ndim,
+                                                                  len(self._points)))
+
+        self.bounds_error = bounds_error
+        self.method = method
+        self.fill_value = fill_value
+
+    @property
+    def points(self):
```
Comment (pllim) [inline astropy/modeling/tabular.py]: Need a docstring here to document this property.

File: astropy/modeling/tabular.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+    if name is None:
+        model_id = Tabular._id
+        Tabular._id += 1
+        name = 'Tabular{0}'.format(model_id)
+
+    return type(str(name), (Tabular,), members)
+
+
+Tabular1D = tabular_model(1, name='Tabular1D')
+Tabular2D = tabular_model(2, name='Tabular2D')
```
Comment (pllim) [inline astropy/modeling/tabular.py]: Need docstrings for both `Tabular1D` and `Tabular2D`.

File: astropy/modeling/tests/test_models.py (hunk 4 above)
Comment (nden) [inline astropy/modeling/tests/test_models.py]: good catch, I'll change it.

File: astropy/modeling/tests/test_models.py (hunk 5 above)
Comment (nden) [inline astropy/modeling/tests/test_models.py]: They are equivalent. This actually exercises the code, so it doesn't really matter.

File: astropy/modeling/tests/test_models.py (hunk 7 above)
Comment (nden) [inline astropy/modeling/tests/test_models.py]: removed it.

File: astropy/modeling/tabular.py (hunk 3 above)
Comment (nden) [inline astropy/modeling/tabular.py]: `scipy.interpn` parameter `points`  takes a tuple in all cases. In 1D it takes e.g. `points=([1,2,3], )`. This seemed unintuitive to me so I made it be just an iterative and turn it into a tuple internally. So using  `squeeze` is not correct. The question is should `Tabular` accept a tuple for `points` in the 1D case or should it be left as it is.

**Comment by nden** [comment]
@pllim I addressed most if not all of your comments.

File: astropy/modeling/tabular.py (hunk 3 above)
Comment (pllim) [inline astropy/modeling/tabular.py]: I think converting it to a format for `interpn` should not be exposed publicly as `self.points`. This is because someone might decide to subclass `Tabular` and define their own way of interpolation using a completely different API. It would be messy if they have to re-convert the points from this special kind of tuple back to normal arrays for their usage.

tl;dr -- Public-facing `points` should be generic. Points re-formatted for `interpn` should be private and not affect subclass overwritting `evaluate()`.

**Comment by pllim** [comment]
Failed tests caused by Sphinx cannot find the docstrings

```python
.../docs/modeling/index.rst:3: WARNING: py:obj reference target not found: Tabular1D
.../docs/modeling/index.rst:3: WARNING: py:obj reference target not found: Tabular2D
```

and tests running even when SciPy is not installed

```python
ImportError: This model requires scipy >= v0.14
```

File: astropy/modeling/tabular.py [hunk 10]
```diff
@@ -94,7 +94,11 @@ class Tabular(Model):
     def __init__(self, points=None, lookup_table=None, method='linear',
                  bounds_error=True, fill_value=np.nan, **kwargs):
 
+        n_models = kwargs.get('n_models', 1)
+        if n_models > 1:
```
Comment (pllim) [inline astropy/modeling/tabular.py]: Either add a test for this `if` block or mark it as `# pragma: no cover`

File: astropy/modeling/tabular.py [hunk 11]
```diff
@@ -120,8 +124,33 @@ def __init__(self, points=None, lookup_table=None, method='linear',
...
+        """
         return self._points
 
+    @property
+    def bounding_box(self):
+        """
+        Tuple defining the default ``bounding_box`` limits,
+        ``(points_low, points_high)``.
+
+        Examples
```
Comment (pllim) [inline astropy/modeling/tabular.py]: Just want to make sure: The example in this docstring is tested, right? Otherwise, we need to add a test for this.

File: astropy/modeling/tabular.py [hunk 12]
```diff
@@ -190,4 +219,62 @@ def tabular_model(dim, name=None):
...
+    ImportError
+        Scipy is not installed.
+
+    Notes
+    -----
+    Uses `scipy.interpolate.interpn`.
+"""
+
+
+Tabular1D.__doc__  =  """
```
Comment (pllim) [inline astropy/modeling/tabular.py]: For some reason, the Sphinx test on Travis is not picking this up. Does it work when you compile the doc locally?

Also, why not also use `+ _tab_docs` for `Tabular` base class itself?

File: astropy/modeling/tests/test_models.py [hunk 13]
```diff
@@ -33,7 +33,9 @@
     HAS_SCIPY = False
 
 if HAS_SCIPY and not minversion(scipy, "0.14"):
-    HAS_SCIPY = False
+    HAS_SCIPY_14 = False
+else:
+    HAS_SCIPY_14 = True
```
Comment (pllim) [inline astropy/modeling/tests/test_models.py]: The correct logic here is:

```python
else:
    HAS_SCIPY_14 = HAS_SCIPY
```

File: astropy/modeling/tests/test_models.py [hunk 14]
```diff
@@ -23,14 +23,20 @@
...
+    import scipy
     from scipy import optimize  # pylint: disable=W0611
     HAS_SCIPY = True
 except ImportError:
     HAS_SCIPY = False
 
+if HAS_SCIPY and not minversion(scipy, "0.14"):
+    HAS_SCIPY_14 = False
+else:
+    HAS_SCIPY_14 = True
```
Comment (pllim) [inline astropy/modeling/tests/test_models.py]:
```python
HAS_SCIPY_14 = HAS_SCIPY
```

File: astropy/modeling/tabular.py [hunk 15]
```diff
@@ -0,0 +1,280 @@
...
+            Input coordinates. The number of inputs must be equal
+            to the dimensions of the lookup table.
+        """
+        inputs = np.array(inputs[: self.n_inputs]).T
+        if not has_scipy:
+            raise ImportError("This model requires scipy >= v0.14")
+        return interpn(self.points, self.lookup_table, inputs,
+                       method=self.method, bounds_error=self.bounds_error,
+                       fill_value=self.fill_value)
+
```
Comment (pllim) [inline astropy/modeling/tabular.py]: Add `__repr__()` and `__str__()` to expose `points` and `lookup_table` in info printouts/strings? (Please see my prior comment on this matter.)

File: astropy/modeling/tabular.py (hunk 12 above)
Comment (pllim) [inline astropy/modeling/tabular.py]: Also, if you are so inclined, there is also [InheritDocstrings](http://docs.astropy.org/en/latest/api/astropy.utils.misc.InheritDocstrings.html) that you can use.

**Comment by nden** [comment]
@pllim Tests and docs passing now. It's ready for final review.

**Comment by pllim** [comment]
@nden , hopefully in a few days. Thank you!

File: astropy/modeling/tabular.py [hunk 16]
```diff
@@ -0,0 +1,309 @@
...
+"""
+
+from __future__ import (absolute_import, unicode_literals, division,
+                        print_function)
+import abc
+import numpy as np
+from .core import Model
+from ..utils import minversion
+
+has_scipy = False
```
Comment (pllim) [inline astropy/modeling/tabular.py]: Is this necessary? `has_scipy` is set in the `try-except` block anyway.

File: astropy/modeling/tabular.py (hunk 3 above)
Comment (pllim) [inline astropy/modeling/tabular.py]: @nden, so the `points` remain a tuple? We talked about it but I don't recall any specific conclusion.

Comment (pllim) [inline astropy/modeling/tabular.py]: Never mind. It appears to be array now. Please disregard my question above, and thanks!

File: astropy/modeling/tabular.py (hunk 16 above)
Comment (pllim) [inline astropy/modeling/tabular.py]: I think this is the last remaining question from me. :sweat_smile:

**Comment by pllim** [comment]
@nden, one final request before merge. Is it possible to squash all the commits into one? No point to merge with commits like "fixing test" or "fixing typo". Thanks!

**Comment by pllim** [comment]
Thanks!!! :tada:
